### PR TITLE
feat: Add essential-node-types crate with BigBangBlock type

### LINKED ISSUE #52: Create BigBangBlock type
Add a big bang block type like:

```rust
pub struct BigBangBlock {
    pub block_state_address: ContentAddress,
    pub contract_registry_address: ContentAddress,
    pub solution: Solution,
}
```

Example YAML:

```yaml
block-state-address: 0x...
contract-registry-address: 0x...
solution: <serialized-solution>
```

## To-Do

- [x] Add the `essential-node-types` crate.
- [x] Add `BigBangBlock` to `essential-node-types`.
- [x] Add default `BigBangBlock` config as `big-bang-block.yml` to repo.
- [x] For now, `include_str!` the `big-bang-block.yml` in `essential-node-types`.
- [x] Add `BigBangBlock` to the node configuration.
- [x] In node run, if no blocks, apply big bang block. This creates and inserts a block at timestamp `0`.
- [x] Add CLI parameter that accepts path to serialized big bang block file in yaml or json.
- [x] Update validation to check block `0` against `BigBangBlock`?
- [ ] RM  old predicate and contract DB interactions. Includes contract, predicate and contract_predicate tables.
- [ ] Change `update_state` sql to use a contract CA directly (rather than ID) or just make it use a key.
- [ ] Expose `&Contract` -> `Vec<Mutation>` fn - from `essential-node-types`?

### PR DESCRIPTION
**Pull request by mitchmindtree**
Closes #52.

Partially addresses #77.

### CODE REVIEW

File: crates/node/src/validate.rs [hunk 1]
```diff
@@ -85,8 +83,8 @@ pub async fn validate_solution(
             .expect("time must be valid"),
         solutions: vec![solution],
     };
-    drop(tx);
-    validate(conn_pool, &block).await
+    tx.commit()?;
```
Comment (mitchmindtree) [inline crates/node/src/validate.rs:86]: UNDO

File: crates/node/src/validate.rs [hunk 2]
```diff
@@ -153,12 +112,17 @@ pub async fn validate(
...
-        let get_predicate = |addr: &PredicateAddress| {
-            Arc::clone(
-                predicates
-                    .get(addr)
-                    .expect("predicate must have been fetched in the previous step"),
-            )
+
+        // Create the `predicates` map.
+        let predicates: Arc<HashMap<PredicateAddress, Arc<Predicate>>> = Arc::new(
+            query_solution_predicates(&post_state, contract_registry, &solution.data).await?,
```
Comment (mitchmindtree) [inline crates/node/src/validate.rs:118]: Don't return early on missing predicate

Comment (mitchmindtree) [reply crates/node/src/validate.rs:118]: Use outcome

File: crates/node/src/validate.rs [hunk 3]
```diff
@@ -264,6 +218,116 @@ impl StateRead for State {
...
+            return Err(SolutionPredicatesError::MissingPredicate(pred_addr.clone()));
+        };
+        predicates.insert(pred_addr, Arc::new(pred));
+    }
+    Ok(predicates)
+}
+
+/// Query for the predicate with the given address within state.
+// TODO: Take a connection pool and perform these queries in parallel.
+// #[cfg_attr(feature = "tracing", tracing::instrument())]
```
Comment (mitchmindtree) [inline crates/node/src/validate.rs:251]: rm

File: crates/node/src/validate.rs [hunk 4]
```diff
@@ -264,6 +218,116 @@ impl StateRead for State {
...
+#[cfg_attr(feature = "tracing", tracing::instrument(skip_all, err))]
+fn query_predicate(
+    conn: &rusqlite::Connection,
+    contract_registry: &ContentAddress,
+    pred_addr: &PredicateAddress,
+    block_number: Word,
+    solution_ix: u64,
+    pre_state: bool,
```
Comment (mitchmindtree) [inline crates/node/src/validate.rs:259]: get rid of this, and use `false` directly as we always want post state

File: crates/node/src/validation/tests.rs [hunk 5]
```diff
@@ -106,56 +100,58 @@ async fn test_invalid_block_validation() {
...
-    insert_block_and_send_notification(&mut conn, &blocks[1], &block_tx);
-    tokio::time::sleep(Duration::from_millis(100)).await;
-    // Assert validation progress is block 1
-    assert_validation_progress_is_some(&conn, &hashes[1]);
-
-    handle.close().await.unwrap();
-}
+// NOTE: Temporarily disabled until issue `#100` is resolved as
+//       this test requires the ability to query non-finalized blocks.
+// #[tokio::test]
```
Comment (mitchmindtree) [inline crates/node/src/validation/tests.rs:105]: ignore not comment

**Review by freesig — APPROVED** [review]
Epic

**Comment by supiket** [comment]
bling blang blonck

**Pull request merged**